### ISSUE TITLE
build-style/go.sh: please move part of do_build() to post_extract() for nonstandard builds.

### ISSUE THREAD

**Issue by ackalker** [voidlinux/void-packages#5816]
To explain: after regular setup (`GOPATH` etc.), `github.com/docker/runc` needs to be built by running `make` from its source directory, or else the build will Git clone and import the entire Docker ecosystem and then some (and finally fail).
See: https://github.com/docker/docker/blob/v1.13.1/hack/dockerfile/install-binaries.sh#L23-L26

I'm referring specifically to these lines: https://github.com/voidlinux/void-packages/blob/1c977b0ba08bb50b6040d0f5af51cda07c4b4b33/common/build-style/go.sh#L6-L12

I think these lines technically don't belong to the `do_build` step itself, they set up the import directory and `go_package` default value.
I suppose it would be better to place them in a a `post-extract` hook. I do realize that that would mean the hook gets run for every package (and should bail immediately if the package is not a Go buildstyle package), but currently the only alternative is to override `do_build()` with most of its original contents from `build-style/go.sh` and add the non-standard build commands, `make` in this case.

Any comments on this are very welcome.

**Comment by ebfe** [comment]
If the build system is `make`, why do you want to use `build_style=go`?

**Comment by ackalker** [comment]
@ebfe Because `go` build style also sets up and exports environment variables specifically for use by `go build`, Also, the `gnu-makefile` build-style exports a conflicting `LDFLAGS` which interferes with yet another package i'm creating (Go uses `CGO_LDFLAGS`).
See `common/environment/build-style/go.sh`.

Also, I think that it is still possible that `docker` could become a `go get`able package like the others...

**Comment by ackalker** [comment]
To be clear: `github.com/docker/runc` _is_ a Go package, it is just not yet properly vendored in and requires special treatment. (it is an forked, unreleased, commit-pinned version of `github.com/opencontainers/runc` required by Docker 1.13.1, don't ask me, I'm not a Docker dev...).

**Comment by ackalker** [comment]
Working package template, illustrating the issue:
```sh
# Template file for 'docker-runc'
pkgname=docker-runc
version=1.0.0.rc2.20170129
revision=1
_githash=9df8b306d01f59d3a8029be411de015b7304dd8f
wrksrc="runc-${_githash}"
build_style=go
go_import_path=github.com/docker/runc
hostmakedepends="git pkg-config"
makedepends="libseccomp-devel"
short_desc="runc container cli tools"
maintainer="Alain Kalker <a.c.kalker@gmail.com>"
license="Apache-2.0"
homepage="https://github.com/docker/runc"
distfiles="https://github.com/docker/runc/archive/${_githash}.tar.gz"
checksum=11f69923a786efb82b65d9f1eec0ac665a985e631a67f3f4d466dac720efacdd

do_build() {
	local path="${GOPATH}/src/${go_import_path}"
	mkdir -p "$(dirname ${path})"
	ln -fs $PWD "${path}"
	make -C "${path}" BUILDTAGS='seccomp'
}
do_install() {
	vbin "${GOPATH}/src/${go_import_path}/runc" docker-runc
}
```
Sorry, I should have posted this earlier, but it wasn't finished.
(Yes, the `do_install()` fixes yet another unusual pattern in this Go package. It is definitely not standard fare.)

**Comment by ebfe** [comment]
Ok. I think the build_style shouldn't define pre_*/post_* functions as these are used to amend build_style behaviour. Before moving the paths stuff to do_extract one would need to review existing go templates to ensure this doesn't break expectations elsewhere.

**Comment by ackalker** [comment]
Fair enough. When I have some more time, I would like to help with that.

**Comment by ebfe** [comment]
Should be resolved with 92d6e1a521871726151d22b2c1dc1b74efd5f290
